networking/p2p/src/integration-test/java/tech/pegasys/artemis/networking/p2p/jvmlibp2p/BeaconBlocksByRangeIntegrationTest.java

```diff
@@ -0,0 +1,127 @@
+import static tech.pegasys.artemis.network.p2p.jvmlibp2p.ChainStorageClientFactory.initChainStorageClient;
+import static tech.pegasys.artemis.util.Waiter.waitFor;
+
+import com.google.common.eventbus.EventBus;
+import com.google.common.primitives.UnsignedLong;
+import java.util.ArrayList;
+import java.util.List;
+import org.apache.tuweni.bytes.Bytes32;
+import org.junit.jupiter.api.AfterEach;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import tech.pegasys.artemis.datastructures.blocks.BeaconBlock;
+import tech.pegasys.artemis.datastructures.state.BeaconState;
+import tech.pegasys.artemis.datastructures.util.DataStructureUtil;
+import tech.pegasys.artemis.network.p2p.jvmlibp2p.NetworkFactory;
+import tech.pegasys.artemis.networking.p2p.JvmLibP2PNetwork;
+import tech.pegasys.artemis.storage.ChainStorageClient;
+import tech.pegasys.artemis.storage.Store.Transaction;
+
+public class BeaconBlocksByRangeIntegrationTest {
+
+  private final NetworkFactory networkFactory = new NetworkFactory();
+  private int seed = 1000;
+  private Peer peer1;
+  private ChainStorageClient storageClient1;
+
+  @BeforeEach
+  public void setUp() throws Exception {
+    final EventBus eventBus1 = new EventBus();
+    storageClient1 = new ChainStorageClient(eventBus1);
+    final JvmLibP2PNetwork network1 = networkFactory.startNetwork(eventBus1, storageClient1);
+    final JvmLibP2PNetwork network2 = networkFactory.startNetwork(network1);
+    peer1 = network2.getPeerManager().getAvailablePeer(network1.getPeerId()).orElseThrow();
+  }
+
+  @AfterEach
+  public void tearDown() {
+    networkFactory.stopAll();
+  }
+
+  @Test
+  public void shouldSendEmptyResponsePreGenesisEvent() throws Exception {
+    final List<BeaconBlock> response = requestBlocks(Bytes32.ZERO);
+    assertThat(response).isEmpty();
+  }
+
+  @Test
+  public void shouldSendEmptyResponseWhenNoBlocksAreAvailable() throws Exception {
+    initChainStorageClient(storageClient1);
+    final List<BeaconBlock> response = requestBlocks(storageClient1.getBestBlockRoot());
+    assertThat(response).isEmpty();
+  }
+
+  @Test
+  public void shouldSendEmptyResponseWhenHeadBlockRootDoesNotMatchAnyBlock() throws Exception {
+    initChainStorageClient(storageClient1);
+    final List<BeaconBlock> response = requestBlocks(Bytes32.ZERO);
+    assertThat(response).isEmpty();
+  }
+
+  @Test
+  public void shouldSendEmptyResponseWhenHeadBlockRootIsNotOnCanonicalChain() throws Exception {
+    initChainStorageClient(storageClient1);
+    final BeaconBlock nonCanonicalBlock = addBlock(1);
+    final List<BeaconBlock> response = requestBlocks(nonCanonicalBlock.hash_tree_root());
+    assertThat(response).isEmpty();
+  }
+
+  @Test
+  public void shouldRespondWithBlocksFromCanonicalChain() throws Exception {
+    initChainStorageClient(storageClient1);
+    final BeaconBlock block1 = addBlock(1);
+    final BeaconBlock block2 = addBlock(2);
```

Comment:
The blocks created by `createBlockAtSlot` aren't entirely valid (randao reveal is wrong) and it's not actually processing the block to create the post state.  Going through the exercise did however reveal a couple of bugs so it's definitely worth hooking up a real block producer.

Will put this PR on hold for a bit while I investigate how to build out `BeaconChainUtil` so it can add fully processed blocks into storage for tests as raise that as a standalone PR. May not get something done until next week though.

Reply:
Switched to using `BeaconChainUtil` to generate the blocks and added a function to get it to import them as well.  Promptly discovered a bug because "block root" is actually the signing root, not the hash tree root. Now fixed.